### ISSUE TITLE
Champion "Null-conditional await"

### ISSUE THREAD

**Issue by MadsTorgersen** [dotnet/csharplang#35]
- [X] Proposal added: [proposals/null-conditional-await.md](https://github.com/dotnet/csharplang/blob/master/proposals/null-conditional-await.md)
- [ ] Discussed in LDM
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

**Comment by dmitriyse** [comment]
As I am understand it's port from https://github.com/dotnet/roslyn/issues/7171

**Comment by yaakov-h** [comment]
> awaits e if it is non-null, otherwise it results in `null`

What if I have `Task<T>` where `T` is a struct/scalar, or `Task` without a result type?

**Comment by alrz** [comment]
awaiting Task yields void so you can't assign it. for a value type I believe it'll be T?.

**Comment by gafter** [comment]
Confirming @alrz 's response,

If `e` is of type `Task`, then `await? e;` would do nothing if `e` is `null`, and await `e` if it is not `null`.

If `e` is of type `Task<K>` where `K` is a value type, then `await? e` would yield a value of type `K?`.

**Comment by ljw1004** [comment]
I think this proposal doesn't quit hit the key scenario, for the same reason as we made the last-minute change to "short circuit ?."

Let's review that short-circuit first. We initially had users put ?. all the way through, because we said every single ?. was evaluated from left to right. If `x()` returned null, then `x()?.y()` would return null, and so you had to put `?.z` else otherwise it would throw.
```csharp
  x()?.y()?.z
```

But this wasn't so nice... (1) it created "question mark pollution" everywhere, (2) it THREW AWAY legitimate null-checking information -- in the case where `x()` returns non-null, and you want to assert that `y()` will never return null, so you want to write `x()?.y().z`. The left-to-right order threw away your ability to distinguish that.

Let's now look at the proposed `await?` operator. And pretend for a moment that await could be written in a fluent left-to-right syntax, rather than being forced to the left...
```csharp
  await? x()?.y();
  x()?.y()?.<<AWAIT>>;
```

The proposal as it stands requires you to write `await?`. This has the same problems as left-to-write conditional evaluation: it peppers your code with too many awaits, and it throws away the legitimate check you might want to make that y() never returns a null Task.

...

**KEY SCENARIO.** All of that discussion is theoretical. Let's get concrete. The places where I've wanted this feature (and I've wanted it a lot) have almost all been to do with `?.` operators on the right hand side. I reckon this is the key use-case for this scenario:
```csharp
   await? customers.FirstOrDefault()?.loadedTask;
```

**PROPOSAL2:** Let's make the same short-circuit ?. evaluation order apply to `await` as well. So, if the await operand involved a ?. that short-circuited the remainder of the expression, then it can also short-circuit any await operator that operates on the expression.

Written out like that, the proposal feels partly natural but partly irregular/unpredictable. And to avoid the unpredictability, that's why I end up preferring @bbarry's original suggestion:

**PROPOSAL3:** Let's say that `await` always does the null-check, on the grounds that (1) awaiting null feels naturally like a no-op, (2) using the await operator is a kind of goofy way to put in a "non-null" assertion into your code, and if you really wanted a non-null check, then there are much better ways to write it.


Note: I've spent the past two months immersed in the Flow language. I really like how it lets me write clean code with a minimum of annotations and it figures them all out for me. I guess that `await?` feels extra busy in the light of that experience.

**Comment by gafter** [comment]
@ljw1004 But isn't your proposal an incompatible change? What if someone has come to depend on the code throwing a `NullReferenceException`? `</joking>`

**Comment by alrz** [comment]
Since `return null` would still create a `Task<T>` in an `async` method I also think making `await` a no op in case the expression is null is ok. However, it should not *always* check for null, only when it's being used with null-conditional member access. Though that *might* not be always the case.

**Comment by svick** [comment]
@alrz 

> However, it should not *always* check for null, only when it's being used with null-conditional member access.

I would find it confusing if `await a?.b;` gave different result than `var task = a?.b; await task;`. Something similarly confusing already happens with `?.` followed by `.`, but in that case there is a good reason for it. What is the reason here?

**Comment by alrz** [comment]
I was pointing out an issue that would arise if we want to deliberately avoid `await?` syntax. I think it's not "extra busy" as @ljw1004 claims. And I don't think the compile should "always" emit null check when we're using `await` to make awaiting a null a no op.

**Comment by gafter** [comment]
@ljw1004 I think there is a compatibility issue with your proposal.

If we add the feature that `await` automatically does the null check (whether or not that depends on the syntax of the operand), then people will write code depending on that. But that code will also compile cleanly with an earlier version of the compiler, where it will produce code that throws an exception.

I think that is a fatal flaw.

**Comment by bbarry** [comment]
What if said change was tied to a CLR version upgrade (say for default interface methods). A compiler targeting the new CLR could let `await null` be a no-op. and an older compiler wouldn't be able to target the new framework.

**Comment by jnm2** [comment]
@alrz Always checking null is no worse than what every `using` statement does today. It seems like the most natural thing in the world to me.

**Comment by jnm2** [comment]
@gafter Just brainstorming. To solve the old compiler problem, it could be useful to plan to add the capacity to opt in to new compiler behaviors in the csproj SDK which translate to csc.exe switches, in the same vein as `<AllowUnsafeCode>` and `<CheckForOverflowUnderflow>`, but which would cause old compilers to error. Here's what that might look like:


### Default template, allows `await null`

(SDK sets `<AllowAwaitNull>true</AllowAwaitNull>`)

Csproj, does not override `AllowAwaitNull` so it stays true:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net462</TargetFramework>
  </PropertyGroup>
</Project>
```

Class1.cs
```c#
class Class1
{
    public async Task Foo(Bar x) => await x?.WhenX;
}
```

**Old compiler** gets passed `-CompatSwitch:AllowAwaitNull` and is smart enough to refuse to build at all (whether or not you await) because it doesn't recognize the `AllowAwaitNull` switch.

**New compiler** emits the null check because it recognizes the switch.


### In the rare scenario where you need to compile using an older compiler:

(SDK sets `<AllowAwaitNull>true</AllowAwaitNull>`)

Csproj, overrides `AllowAwaitNull` so that it can build against older compilers:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net462</TargetFramework>
    <AllowAwaitNull>false</AllowAwaitNull>
  </PropertyGroup>
</Project>
```

Class1.cs
```c#
class Class1
{
    public async Task Foo(Bar x) => await x?.WhenX;
}
```

**Old compiler** does not get passed `-CompatSwitch:AllowAwaitNull` and acts as it always has, and an analyzer sees the `await x?.y` and warns that you may be awaiting null.

**New compiler** skips the null check because it was not given the switch, and an analyzer sees the `await x?.y` and warns that you may be awaiting null.

**Comment by svick** [comment]
@bbarry @jnm2 I'm not sure this is a problem that can be solved with technical solutions. I think it's an undesirable situation when you find some code e.g. on Stack Overflow, you copy it to your project and it's broken in a subtle way, because you're using a different version of the compiler (compiler error is fine, exception isn't).

**Comment by ljw1004** [comment]
@gafter agreed it's a fatal flaw. That's a shame.

**Comment by jamesqo** [comment]
I wonder if it would make much of a performance difference to return `null` instead of `Task.CompletedTask` for tasks that complete synchronously. I'm working on an app right now that awaits a huge (thousands) amount of tasks, most of which are completed. I could check `!IsCompleted` on each task before awaiting it, but it would feel cleaner to do `await? task`.

**Comment by yaakov-h** [comment]
@jamesqo Isn't that what `ValueTask`is for?

**Comment by jamesqo** [comment]
@yaakov-h `ValueTask<T>` is only available as a counterpart to the generic `Task<T>` type. There is no non-generic `ValueTask`.

**Comment by jamesqo** [comment]
To be fair though, I'm not 100% sure `await? x` will provide much benefit over awaiting a completed task. From what I can see `GetAwaiter()`, and `IsCompleted` and `GetResult()` on that awaiter reduce to just a few flag checks for a completed task. Maybe it wouldn't make so much of a perf difference to have `await?`.

**Comment by Joe4evr** [comment]
> I'm not 100% sure await? x will provide much benefit over awaiting a completed task.

It's more for cases like `var x = await foo?.SomeTask();` If `foo` turns out to be null, then the `await` still attempts to call `((Task<T>)null).GetAwaiter().GetResult()` which turns into a cryptic NRE.

**Comment by jamesqo** [comment]
@Joe4evr Yes, I know; I wasn't criticizing the feature, I meant performance benefit. Sorry for any confusion.

**Comment by gundermanc** [comment]
I had opened [an issue](https://github.com/dotnet/roslyn/issues/22784) in the Roslyn repo to add a warning for this before seeing this thread.

I don't think that the referenced `await?` operator addresses the primary risk of the await operator's current behavior. The lack of a null check for the await operator makes async calls using `?.` a bit of a special case that the developer has to be cognizant of, and the NRE is difficult to diagnose if you haven't seen it before.

Though a breaking behavioral change is probably not possible, having a warning and changing the exception message for this case would at least aid the developer in prevention/diagnosis.

**Comment by taori** [comment]
I'm curious about the state of this - as is, it would not be introduced unless someone of the community implements it. Is this correct? At least i didn't see any visible branches named feature(s)/await.

I would really love this feature to exist though.

**Comment by alrz** [comment]
"Any Time" milestone description:

>We probably wouldn't devote resources to implementing this, but we'd likely accept a PR that cleanly implements a solution.

**Comment by taori** [comment]
@alrz Yes - My question was more about the "I didn't see any visible branches"-part. I've read that description too. I'm just making sure there isn't someone working on this already.

**Comment by Jopie64** [comment]
Hi. Just a thought that came up when reading this thread.
When the [nullable-reference-types](https://blogs.msdn.microsoft.com/dotnet/2017/11/15/nullable-reference-types-in-csharp/) feature is turned on, `await?` should not be necessary anymore. `await` (without ?) could simply return `T` in case of awaiting a `Task<T>` and return a `T?` in case of awaiting a `Task<T>?`. The compiler can detect when used erroneously.

**Comment by Jopie64** [comment]
Sorry, changed my mind. It should error out when using `await` on a `Task<T>?`. Otherwise you cannot express that you don't want to wait on a nullable task when it is a void task or a `Task<T?>`.

**Comment by canton7** [comment]
Since nobody's mentioned it yet:

I think this is relevant for `IAsyncDisposable`. A common pattern is `foo?.Dispose()` - it's a bit counter-intuitive that this pattern doesn't carry across to `await foo?.DisposeAsync()`.

**Comment by yahorsi** [comment]
Any progress on that? It really hurts and initial issue was created back in 2015 (((

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/milestone/14

> We probably wouldn't devote resources to implementing this, but we'd likely accept a PR that cleanly implements a solution.

**Comment by legistek** [comment]
I've been curious about this for awhile. Count me in the category of `await null` should simply do nothing similar to the way `using (null)` does nothing when it comes time to dispose. This way we can do things like `await thing?.DoSomethingAsync()` without having extra lines of null-checking. 

Maybe this proposal is obsolete in the age of nullable reference types though, which I admit I haven't gotten into yet.

**Comment by paulhickman-a365** [comment]
In https://github.com/dotnet/csharplang/issues/35#issuecomment-287893347 ljw1004 started off his discussion with a "consider if await was fluent" scenario.  Is there a reason why this couldn't actually be implemented?

i.e. make the following valid alternatives to await in front of the expression.
````
thing.DoSomethingAsync().await;
thing?.DoSomethingAsync().await;
thing?.DoSomethingAsync()?.await;
var t = thing.DoSomethingAsync().await;
var t = thing?.DoSomethingAsync().await;
var t = thing?.DoSomethingAsync()?.await;
````
This gets around the issue of ````?```` pollution and allows you to express that ````thing```` may be null but ````DoSomethingAsync()```` shouldn't return null if it is executed.  It also means the semantics of the existing syntax of putting await in front of an expression does not change.

It's arguable whether the last example is should be valid or not, as there you do have a method with return type of Task returning null. ````?.await```` could still raise an error whilst accepting ````?.```` operations further to the left in the expression. 

It would also have the benefit outside of null coalescing that it makes it cleaner to chain operations on the results of async calls e.g.

````
var count =thing.DoSomethingAsync().await.Count();
````
Rather than:
````
var count =(await thing.DoSomethingAsync()).Count();
````

If this is considered, I'd also like to see a shortcut for ConfigureAwait(false) by giving an argument to the await - i.e. the following are equivalent:
````
var count =(await thing.DoSomethingAsync().ConfigureAwait(false)).Count();
var count =thing.DoSomethingAsync().ConfigureAwait(false).await.Count();
var count =thing.DoSomethingAsync().await(false).Count();
````

Obviously, it would introduce a conflict with properties/methods called "await" but it would be a compile time breaking change, rather than something that causes a runtime exception / change in runtime behaviour, and await has been a keyword long enough now that I can't believe any developers actually still use it as a property/method name in code that is being actively ported into new language versions, or use it in newly written code.

**Comment by HaloFour** [comment]
@paulhickman-a365 

>  Is there a reason why this couldn't actually be implemented?

The syntax you suggested is already legal in C#, there's nothing stopping some arbitrary expression from having an `await` property.  Sure, the language could consider it as a contextual keyword where such a property doesn't exist, but that would be confusing.  And this only serves to add duplicate syntax for an existing feature which doesn't solve any problems except when it comes to stylistic preferences.

**Comment by gafter** [comment]
@HaloFour No, the suggested syntax is not already legal C#. @paulhickman-a365 's idea is workable.

See https://sharplab.io/#v2:D4AQTAjAsAUCAMACEECsBuWsQGZlkQGFEBvWRC5PAFUQFkAKASkQF4A+RAOwFcAbPphiVE5SrmQAOZADZ6YZmIplhIygH029ZgDoQATiEiAvrFNw84RNVgqREgJZcALsn1tOEIeaA===

**Comment by paulhickman-a365** [comment]
I hadn't realised it was already illegal - I thought I was clobbering existing property names. 

Thinking about it some more, I think ````.await()```` may be preferable to ````.await````. Although it introduces additional punctuation, the act of awaiting is more like invoking a method than accessing a property so I think it is easier to understand what is happening particularly for new developers.

**Comment by theunrepentantgeek** [comment]
Using the syntax `.await()` is already valid C# syntax, all you need to do is to write extension methods on `Task` and `Task<T>` ([Sharplap](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0ATEBqAPgAQCYBGAWACh8AGAAn2IFYBuC1ygZjsJoBUALAJYA7AOYUA3hRrS6nfAA46SGgBEIAZQgBbGABdBogIIBnAJ5CwACgCUNKTMnkZzugE4lAOgCaAmABsMGxYnGQBfe2kI2TpFfCQAHmFdAD4aAGEAQz8wTR19YRETcytrKMcXGXx3OO9fAKCo53wAdhoUQmDncPJuinw5YmUiGiKLAFEEXRghYwEIGYko/rpBuhQ3S3zjJRpdDOMAa1KQ6XKK3f2DjwAlAFchdWK+KHmIW+M/UwaTmm7nJYGyh4bniPGSm0E2zioNSe0Ox2cZwqLQuhxuMGMtz8uk6YQovQ4XFUMC0EDs5CR0Xw6wA6nxpgA5CDXPS3KBCABqWVuMBsZUaMgAbhkoLsDCIaABeGhCGAAd144u+5zFBQ8alyenFoxKHiqyvO+VEAH51RptFqCjqbHrXAaKkaRKaNRbHdbrKb9dZcdI/jIAWsaHTpizdGyhAUuX4eXyfpTnMLRY6pTL5YqCvaXImaAgU46PJlspq3WYLDavT6syKaKY8+LTYWcq7taXdRWBdJswAvOsFBtZJt5FvFGyeu3eqKhf49IA===)).

The `.await` syntax isn't currently legal C# syntax, but would collide with extension properties (#192, [#11159](https://github.com/dotnet/roslyn/issues/11159)).

**Comment by canton7** [comment]
@theunrepentantgeek I'm afraid I'm missing your point. It's not valid to call an `await()` extension method from inside an `async` method, which is the case that's being discussed here.

**Comment by Jopie64** [comment]
@theunrepentantgeek 
`.await()` as an extension method would be different behavior from the current `await` keyword, because the extension method would be a blocking call (it would block the thread). I think there is no way you currently can replicate the behavior of `async/await` with an extension method.

**Comment by huoyaoyuan** [comment]
I'm starting an implementation here: https://github.com/huoyaoyuan/roslyn/tree/features/conditional-await

**Comment by CyrusNajmabadi** [comment]
I'll champion this.

**Comment by huoyaoyuan** [comment]
@CyrusNajmabadi any update on this?

**Comment by CyrusNajmabadi** [comment]
There is no update on this.

**Comment by huoyaoyuan** [comment]
So what does the tag `Needs Implementation` and milestone `Any Time` means? Is it welcome for implementation?

**Comment by CyrusNajmabadi** [comment]
@333fred For info. From my understanding, it means we're not talking this for 10.0, but are not rejecting it. We could take it in some future release.

**Comment by 333fred** [comment]
Any time generally means any time 🙂. We'd accept a community contribution that implements this feature, subject to compiler team review availability. I don't expect that we have time for this in the near term, however, as we're in full swing for bigger C# 10 features at the moment and this is not a priority.

**Comment by AlFasGD** [comment]
> I'm starting an implementation here: https://github.com/huoyaoyuan/roslyn/tree/features/conditional-await

How's progress on the feature doing?

**Comment by huoyaoyuan** [comment]
> How's progress on the feature doing?

It's laid aside until the team approves it and assign a reviewer.

**Comment by 333fred** [comment]
> How's progress on the feature doing?

The feature is in the any time bucket and tagged needs implementation. The team is not going to do anything on it on our own, but we will accept a community contribution.

**Comment by huoyaoyuan** [comment]
@333fred I had completed a basic implementation. Will you assign someone to review it during C# 11 lifetime?

**Comment by 333fred** [comment]
No, but maybe @jaredpar would.

**Comment by jnm2** [comment]
I just thought of one backwards compatibility thing that I can't see around:

```cs
int x = await taskOfInt;
```

Versus
```cs
int? x = await? taskOfInt;
```

**Comment by Xyncgas** [comment]
because ```if(x !=null) await x;``` is what we have to do right now it increases the letters we have to type, it would be fantastic if we can get the feature

**Comment by CyrusNajmabadi** [comment]
@Xyncgas see https://github.com/dotnet/csharplang/issues/35#issuecomment-922345018

**Comment by Xyncgas** [comment]
> @Xyncgas see [#35 (comment)](https://github.com/dotnet/csharplang/issues/35#issuecomment-922345018)

Can I pay you $100 to do it

**Comment by CyrusNajmabadi** [comment]
> Can I pay you $200 to do it, I can pay you before you finish

I appreciate the gesture.  But i'd rather the money go to better causes :)

**Comment by taori** [comment]
@CyrusNajmabadi Are there any guidelines on how a feature like this would be implemented or a branch in your memory which adds a feature (so i can learn from file changes of the PR)? I usually only write analyzers/quickfixes, but i suppose with some example on how to add such a feature i would be capable of doing this. I've cloned dotnet/roslyn and jumped back and forth to get an idea, but i guess if there are any recommended docs to read on this, implementing this feature would be a much more feasible task.

**Comment by CyrusNajmabadi** [comment]
@333fred any thoughts?

**Comment by 333fred** [comment]
@taori I would suggest coming and talking to us on discord, in one of the communities listed here: https://github.com/dotnet/csharplang/blob/main/Communities.md. This issue isn't likely to be a good place for the type of back and forth real time discussion you'd probably want.

**Comment by legistek** [comment]
I agree with those suggesting to simply change `await` to not throw on null, just as with `using`.

The only code that could be negatively impacted, it seems to me, would be code that relied on the throwing of a null ref exception instead of actually checking for null. We shouldn't let past bad practices impede future progress, especially since, as was pointed out, there is much more appetite for breaking changes now than there was several years ago, and language versions are opt-in anyway. 

FWIW, I would change `foreach` in the same way.

**Comment by jnm2** [comment]
@legistek Changing `await` to not throw on null isn't feasible when awaiting `Task<int>` because you'd need to get an `int?` out of that, and this would change the meaning of existing code such as:

```cs
var someInt = await foo.SomeTask;
// someInt is currently `int`, not `int?`
```

**Comment by AlFasGD** [comment]
I largely agree to this, you cannot break existing behavior that much. `await` was initially designed following a condition that was taken when designing `foreach`; ensure that the thing is not `null`, and throw if it is. Such a design has great advantages, and is the healthiest possible. You cannot simply assume that in any case, the awaited task is null, requiring that you handle the result being null too. In the case of `foreach`, not much changes if you ignore the entire iteration if the iterated collection is null.

We have already crossed that path. The `foreach` and `await` account for nullability. Silently changing `foreach` behavior will only impact bad practices, like relying on the thrown exception. This will, however bring a slight inconsistency to the table, with `await` no longer taking a similar path. The healthiest addition would be `foreach?` and `await?`, which explicitly indicate the implementation detail.

**Comment by legistek** [comment]
> @legistek Changing `await` to not throw on null isn't feasible when awaiting `Task<int>` because you'd need to get an `int?` out of that, and this would change the meaning of existing code such as:
> 
> ```cs
> var someInt = await foo.SomeTaskOfInt;
> // someInt is currently `int`, not `int?`
> ```
> 
> Would every await of a value-typed result start returning `Nullable<T>`? Almost all the time, this would not even be helpful, let alone how confusing of a change in meaning this would be in the language.
> 
> I don't think replacing throwing with using `default(int)` is a good idea, either.

Good point, I hadn't thought of returning non-nullable value types.

My first response would have been to go with `default`, but I do see how that would be a pretty significant behavior change.

**Comment by CyrusNajmabadi** [comment]
> I agree with those suggesting to simply change await to not throw on null, just as with using.

> The only code that could be negatively impacted, it seems to me, would be code that relied on the throwing of a null ref exception instead of actually checking for null.

I disagree with this, for the same reason I think that `.` should continue to throw with a null receiver. 

Yes, it will almost always throw when it encounters a bug in the code.  But that's a virtue. It tells me when I've done something very wrong. Nulls should not be papered over.  They should either be explicitly expected, or they should fail catastrophically. Silently ignoring them is just a way to lead to even worse bugs happening

**Comment by legistek** [comment]
> Nulls should not be papered over.

Ok but what's the logic for allowing `dispose(null)` then?

> await was initially designed following a condition that was taken when designing foreach; ensure that the thing is not null, and throw if it is. Such a design has great advantages, and is the healthiest possible. You cannot simply assume that in any case, the awaited task is null, requiring that you handle the result being null too. 

This is true but it's also worth pointing out that `foreach` and (I believe?) `await` predate `?.`. 

Before `?.`, if you said `await foo.bar()`, you had to check `foo` first or were rolling the dice. So there was no legitimate reason to ever be `await`ing a null value.

But with `?.` the temptation is to write `await foo?.bar()` to skip the null check. This is what I (for one) want to be able to do.

Regarding the value type question, I assume if we used `await?` with a value type, would the resulting type would be nullable right? 

In other words assume `myClass.foo()` returns a `Task<int>`.  If I said:
`var result = await? myClass?.foo()`
`result` would be an `int?` right?

**Comment by CyrusNajmabadi** [comment]
> Ok but what's the logic for allowing using (null) then?

Legacy.  Doing it again, i would absolutely not do that.

**Comment by CyrusNajmabadi** [comment]
`  (I believe?) await predate ?..

Sure.  But it postdates our belief that null-tolerance is not a virtue, which is why alter versions of the language embraced explicit null-handling as the mechanism to use, not being null-tolerant.

**Comment by CyrusNajmabadi** [comment]
> But with ?. the temptation is to write await foo?.bar() to skip the null check. This is what I (for one) want to be able to do.

Sure.  But i don't.  And i'd much rather just say: write `await? foo?.bar()` if your preference is to be null-resilient for tasks.

**Comment by legistek** [comment]
Yeah ok. Well like I said I can live with `await?` as I hadn't considered the nullable value type issue, and it sounds like `await?` solves that then.

I still would let `foreach` blow past the null, but I guess I just like to live dangerously.

>> Ok but what's the logic for allowing using (null) then?

> Legacy. Doing it again, i would absolutely not do that.

Interesting! A true believer then. :)

**Comment by SebastianSchumann** [comment]
In the meantime we have .Net 8.
`ConfigureAwait` has been implemented with `ConfigureAwaitOptions`. If you specify `SuppressThrowing`, an `ArgumentOutOfRangeException` is thrown if the task is not terminated correctly and should return a value.

Wouldn't that also be an option for this `await` implementation?

**Comment by Rekkonnect** [comment]
What you're saying is that the nullability annotation of the type should control whether the task is allowed to be null and thus suppress throwing the null exception. Keep in mind that `Task` is a reference type, and the annotation is part of the metadata, not within the actual type system. The awaited literal's type `(Type?)null` is still a `null` used in a place where `Task` (or any implementing an async iterator) is expected. During runtime you have no idea about the type being converted into `Type?`, this would be entirely a C# detail implemented in a very hacky way, in my opinion.

**Comment by SebastianSchumann** [comment]
> What you're saying is that the nullability annotation of the type should control whether the task is allowed to be null and thus suppress throwing the null exception. 

No. My example code should only illustrate how `await` might work. I am aware that the nullable annotation is only metadata and you forward a null reference. 

But the compiler "knows" whether the task has a return value or not. If not no exception should be thrown. If the result value is used in any way (assignment to variable, forward to using, ...) an `ArgumentOutOfRangeException` should be thrown in the same way like `ConfigureAwait(ConfigureAwaitOptions.SuppressThrowing)` will do.

I edited my sample code to be more clear.

**Comment by alrz** [comment]
To my surprise there are many hits for await on conditional member access most of which are invalid? 

https://github.com/search?q=%2Fawait%5Cs%2B%5Cw%2B%5C%3F%5C.%5Cw%2B%2F%20lang%3AC%23&type=code (not a comprehend query)

Would it make sense if the compiler adds a null check in the presence of a null-conditional access with the current syntax? It seems like that's sort of the expected behavior given that many hits.

**Comment by julealgon** [comment]
It's interesting this popped up again as I'm in the middle of converting large amounts of APIs from sync to async in one of our largest solutions.

I must say this is super dangerous when performing such conversions. I've hit a few cases here where we had "callback/event" properties typed as `Action` that I was forced to convert to `Func<CancellationToken, Task>` to make sure those callbacks were awaitable and could then call async methods inside them.

This caused code to change from this:
```csharp
myClass.SomeCallback?.Invoke();
```

To this:
```csharp
if (myClass.SomeCallback is not null and var someCallback)
{
    await someCallback.Invoke(cancellationToken);
}
```

But that was after I had made the changes and before sending the PR for review that I noticed there could be a problem and switched to the explicit defensive check.

It would've been nice if "it just worked" and I could keep it like:
```csharp
await someCallback?.Invoke(cancellationToken);
```

Which is exactly what @alrz is proposing:
> Would it make sense if the compiler adds a null check in the presence of a null-conditional access with the current syntax? It seems like that's sort of the expected behavior given that many hits.

I'd even go as far as to suggest this null check should be done _regardless_ of the presence of null coalescing operators in the right-hand side. It would make `await` behave similar to how `using` behaves in the presence of `null`. This works just fine, for example, with not even a compiler warning or hint anywhere:

```csharp
using ((IDisposable?)null)
{
    Console.WriteLine("Hello world");
}
```

If `await` behaved the same way I think it would be consistent and not surprising.
